### ISSUE TITLE
Python Overhead

### ISSUE THREAD

**Issue by ryanmrichard** [NWChemEx/NWChemEx#85]
TL;DR our Python API is very heavy, and it gets heavier under MPI.

I'm working on what I think is the first end-to-end run of our code under realistic conditions (through the Python API, and relying on plugins). For the NWX runs I'm running a 1B-MBE of an n-water cluster at the HF/STO-3G level of theory (i.e. n independent water calculations). The MBE module is MPI parallelized and for a 1B-MBE should exhibit almost perfect weak scaling (for water n, using n processes, the time should be the same as a single water molecule). Under the hood of NWX I'm calling PySCF.  In the following, a "straight" call means running something like python3 run_pyscf.py and an "under MPI" call means running something like  mpirun -n x python3 run_pyscf.py .  Times are all wall times measured on my workstation (which is a pretty hefty beast 20 cores, 128 GB RAM). What's important here are the trends and not the literal times themselves, i.e., this is profiling and not meant to be actual timing runs. 

First the baseline, a straight call to PySCF (through their API, not using NWX at all) for one water molecule (no MBE, this is just HF/STO-3G). The entire run takes 0.35 seconds with 0.047 s being the actual SCF. Calling PySCF directly under MPI does not change these times in any appreciable manner for 1 or 2 processes.

Now the baseline for NWX. A straight call to NWX for one water molecule (still going through the MBE code) takes about 16 s with about 4 s being in the MBE. Running it under MPI with one process results in a total time of about 100 s and 50 s of that being spent in the MBE. MPI with two processes results in a total time of about 116s  with 58 s being in the MBE.

1B-MBE on water dimer for NWX. Straight call: total 16 s,  5 s in MBE. Under MPI with one process: total 102 s, MBE 52 s. Under MPI with two processes: total 119 s, MBE 61 s.
1B-MBE on water trimer for NWX. Straight call: total 16s, 5s in MBE. Under MPI with one process: total 102s, MBE 53s. Under MPI with two processes: total 121 s, MBE 64s.

At this point it seems that we've got a rather large latency which is dwarfing the computations. 

Going back to a single water molecule, a straight call to NWX, and dumping the profile info of each module shows that C++ measured the MBE time as about 2.9 s and 2.1s occurs in calling PySCF. The previously reported MBE times were measured from Python, so that suggests that most of the 4s is not actually computational work by NWX or PySCF. (FWIW the total wall time remains about the same even when we dump the profile info).

Running a single water molecule, with NWX under MPI, and dumping the profile info now shows C++ measuring 33.8 s are spent in the MBE, 20.6s of that are in the call to PySCF, and 12.9s are in the call to make basis set (which is a lambda module in my original input script), i.e. 33.5 s appears to be Python overhead.

My `run_pyscf.py` script:
```.py
import pyscf
import sys
import time


def main():
    # Syntax is python run_pyscf.py <number_of_waters>
    assert(len(sys.argv) == 2)

    n_h2o = int(sys.argv[1])  # Number of waters
    dz    = 3.0               # Distance along Z-axis to translate each water
    geom = ''
    for i in range(n_h2o):
        z = str(i * dz)
        geom += 'O  0.0               -0.143222342980786 ' + z + '\n'
        geom += 'H  1.638033502034240  1.136556880358410 ' + z + '\n'
        geom += 'H -1.638033502034240  1.136556880358410 ' + z + '\n'
    
    mol = pyscf.gto.M(atom=geom, unit='B', basis='sto-3g')
    
    start_time = time.time()
    rhf = pyscf.scf.RHF(mol)
    egy = rhf.kernel()
    end_time = time.time()
    print("Elapsed Time: {}".format(end_time - start_time))
    print(egy)

if __name__ == '__main__':
    main()
```

My `run_nwx.py` script:

```.py
import sys

sys.path.append('/home/ryan/workspaces/NWX/NWChemEx/build/Python')
sys.path.append('/home/ryan/workspaces/NWX/ghostfragment/build/Python')
sys.path.append('/home/ryan/workspaces/NWX/FriendZone/src')

from nwchemex import nwchemex, simde, chemist, pluginplay
from ghostfragment import ghostfragment
from friendzone import pyscf
from cppyy.gbl.std import make_shared
import time

class MySTO3G(pluginplay.ModuleBase):
    def __init__(self):
        super().__init__(self)
        self.satisfies_property_type[simde.AtomicOrbitals]()

    def run_(self, inputs, submods):
        [mol] = simde.AtomicOrbitals.unwrap_inputs(inputs)
        aos = nwchemex.apply_basis("sto-3g", mol)
        rv = self.results()
        return simde.AtomicOrbitals.wrap_results(rv, aos)
def water_nmer(n, dz = 3.0):

    mol = chemist.Molecule()

    zO = chemist.Atom.AtomicNumber(8)
    zH = chemist.Atom.AtomicNumber(1)

    for i in range(n):
        z = i * dz
        rO0 = chemist.Atom.coord_type([0.00000000000000, -0.07579039945857, z])
        rH0 = chemist.Atom.coord_type([0.86681456860648, 0.60144316994806, z])
        rH1 = chemist.Atom.coord_type([-0.86681456860648, 0.60144316994806, z])
        mol.push_back(chemist.Atom("O", zO, rO0))
        mol.push_back(chemist.Atom("H", zH, rH0))
        mol.push_back(chemist.Atom("H", zH, rH1))

    return mol


def main():
    assert(len(sys.argv) == 2)
    n = int(sys.argv[1])

    mm = pluginplay.ModuleManager()
    ghostfragment.load_modules(mm)
    pyscf.load_modules(mm)
    mm.add_module("STO-3G", make_shared[MySTO3G]())
    mm.change_submod("FragmentedSystem Driver", "Cap Basis", "STO-3G")
    mm.change_submod("Fragment Based Method", "energy method", "PySCF RHF")
    mod = mm.at("Fragment Based Method")
    mol = water_nmer(n)
    mol_sys = chemist.ChemicalSystem(mol, n * 10)
    aos  = nwchemex.apply_basis("sto-3g", mol)
    mod.change_input["unsigned int"]("GMBE truncation order", 1)

    start_time = time.time()
    [e] = mod.run_as[simde.AOEnergy](aos, mol_sys)
    end_time = time.time()
    print("Time (s): {}".format(end_time - start_time))
    #PySCF energy of 1 water: 
    corr_egy = -74.94208005769586 * n
    print("Actual egy: {} Corr egy: {}".format(e, corr_egy))
    print(mod.profile_info())

if __name__ == '__main__':
    main()
```
(Disclaimer: If you run these scripts you'll get slightly different energies from PySCF vs. NWChemEx, I think that's because I accidentally did the NWX waters in Angstrom vs. Bohr).
